doc/Language/iterating.pod6

```diff
@@ -4,136 +4,321 @@
+
+Raku provides support for both iterables and iterators via the
+L<C<Iterable>|/type/Iterable> and L<C<Iterator>|/type/Iterator> roles.
+
+=head1 The C<Iterable> role
+
+The L<C<Iterable>|/type/Iterable> role is relatively simple and it requires that
+the composed class provides an C<iterator> method, which must return
+an L<C<Iterator>|/type/Iterator> object. The L<C<iterator>|/type/Iterable#method_iterator>
+method is the construct actually used by statements such as C<for>, which
+invokes C<iterator> on the variable it precedes, and then run a block once for
+every item. Similarly other statements, such as array assignment, will make the
+C<Iterable>-composed class behave in the same way.
+
+Any object that can return an C<Iterator> is an iterable. Most container types
+in Raku conform to the C<Iterable> API. To check if a type object or instance
+object composes the C<Iterable> role, and thus implement the
+C<iterator> method, you can use the L<C<does>|/type/Mu#routine_does> method.
+
+=begin code
+say Seq.does(Iterable)  # OUTPUT: «True␤»
+say List.does(Iterable) # OUTPUT: «True␤»
+say Map.does(Iterable)  # OUTPUT: «True␤»
+
+say Str.does(Iterable)  # OUTPUT: «False␤»
+say Int.does(Iterable)  # OUTPUT: «False␤»
+=end code
+
+The following example illustrates the composition of the C<Iterable> role
+into a user-defined C<DNA> class, and the required implementation of the
+C<iterator> method:
 
 =begin code
 class DNA does Iterable {
     has $.chain;
-    method new ($chain where {
-                       $chain ~~ /^^ <[ACGT]>+ $$ / and
-                       $chain.chars %% 3 } ) {
-        self.bless( :$chain );
+    
+    method new(Str $chain where { .match(/^^ <[ACGT]>+ $$/) && .chars %% 3 }) {
+        self.bless(:$chain);
+    }
+
+    method iterator(DNA:D:) { 
+        $.chain.split('', :skip-empty).rotor(3).iterator
     }
+}
+=end code
+
+The C<$.chain.split('', :skip-empty).rotor(3).iterator> statement in the
+C<iterator> method simply splits the DNA strand by the empty string, and group
+the resulting values into lists of trinucleotides. The final result is a
+L<C<Seq>|/type/Seq>, which itself implements the C<Iterable> role and has an
+C<iterator> method that we end up invoking on it.
+
+Instance objects of the C<DNA> class can be iterated over with C<for>, C<map>,
+C<grep>, and array assignment, in the same manner as objects of iterable types,
+such as L<C<Seq>|/type/Seq>, L<C<List>|/type/List> and L<C<Array>|/type/Array>,
+can.
 
-    method iterator(DNA:D:){ $.chain.comb.rotor(3).iterator }
-};
+=begin code :skip-test<compile time error; undeclared name>
+my \dna = DNA.new('ACGTACGTTACT');
 
-my @longer-chain =  DNA.new('ACGTACGTT');
+.join.say for dna;
+# OUTPUT: «ACG␤TAC␤GTT␤ACT»
+
+my @longer-chain = DNA.new('ACGTACGTTACT');
 say @longer-chain.raku;
-# OUTPUT: «[("A", "C", "G"), ("T", "A", "C"), ("G", "T", "T")]␤»
+# OUTPUT: «[("A", "C", "G"), ("T", "A", "C"), ("G", "T", "T"), ("A", "C", "T")]␤»
+=end code
+
```

Comment:
Mentions to "for" and "array assignment" are now further away from the example, so it might be worth the while to explain that in both case "iterator" is being called.